modules/substrate/src/lib.rs

```diff
@@ -14,5 +14,238 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
 
+//! Substrate Bridge Pallet
+//!
+//! This pallet is an on-chain light client for chains which have a notion of finality.
+//!
+//! It has a simple interface for achieving this. First it can import headers to the runtime
+//! storage. During this it will check the validity of the headers and ensure they don't conflict
+//! with any existing headers (e.g they're on a different finalized chain). Secondly it can finalize
+//! an already imported header (and its ancestors) given a valid Grandpa justification.
+//!
+//! With these two functions the pallet is able to form a "source of truth" for what headers have
+//! been finalized on a given Substrate chain. This can be a useful source of info for other
+//! higher-level applications.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+// Runtime-generated enums
+#![allow(clippy::large_enum_variant)]
+
+use crate::verifier::ChainVerifier;
+use bp_substrate::{AuthoritySet, ImportedHeader, ScheduledChange};
+use frame_support::{decl_error, decl_module, decl_storage, dispatch};
+use frame_system::ensure_signed;
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::{marker::PhantomData, prelude::*};
+
 mod justification;
 mod storage_proof;
+mod verifier;
+
+#[cfg(test)]
+mod mock;
+
+type Hash<T> = <T as HeaderT>::Hash;
+type Number<T> = <T as HeaderT>::Number;
+
+pub trait Trait: frame_system::Trait {}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as SubstrateBridge {
+		/// Hash of the best finalized header.
+		BestFinalized: T::Hash;
+		/// Headers which have been imported into the pallet.
+		ImportedHeaders: map hasher(identity) T::Hash => Option<ImportedHeader<T::Header>>;
```

Comment:
My assumption here is that we're bridging a chain that is just like our chain. But you're right, maybe that's not a good assumption since even between Substrate chains we can have differences

Reply:
Let's probably change hasher/block number type that is used in Millau test-chain? E.g. `Hasher` would return `H512` instead of `H256` (we can just concat two `H256` returned by `BlakeTwo256`) and block number e.g. `u128`. This would also help in testing other sub2sub-related pallets as well (`call-dispatch`, `message-lane`) - we'll be sure that our pallets are ready to work with "incompatible" chains? Wdyt?

(not a part of this PR of course)

Reply:
Just for the record, I realised that my previous comment might have been intepreted as endorsing usage of `T::Header` stuff. It's not. I only meant that we can focus on just `sp_runtime_traits::Header`-based headers for now.
We for sure need to decouple source and target types.